Support ECDiffieHellman on X509Certificate2

vcsjones opened the pull request:
This adds `ECDiffieHellman` functionality to `X509Certificate2`. In addition, this also adds arbitrary `SubjectPublicKeyInfo` support for `PublicKey`.

This also makes some under-the-hood changes. Previously, we were always able to know what type of algorithm was used with a given certificate based on the certificate SPKI's OID. In the case of id-ecPublicKey, this is not true anymore. Other plumbing changes include changes to support `ImportFromPem` on `X509Certificate2`.

Work left to be done:

- [x] XML documentation
- [x] Windows PFX tests

Closes #413

vcsjones committed:
Support ECDiffieHellman for GetPublicKey.

vcsjones committed:
Implement GetECDiffieHellmanPrivateKey.

vcsjones committed:
Test changes.

vcsjones committed:
PublicKey tests

vcsjones committed:
Add CopyWithPrivateKey for ECDH.

vcsjones committed:
Fix PEM import for ECDH keys.

vcsjones committed:
Add arbitrary SPKI support for PublicKey.

vcsjones committed:
Test signing PublicKey with CertificateRequest.

vcsjones committed:
Add ECDH to chain tests

vcsjones commented on src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Helpers.cs:
```diff
@@ -242,6 +242,57 @@ public static ReadOnlyMemory<byte> DecodeOctetStringAsMemory(ReadOnlyMemory<byte
             }
         }
+
+        public static bool AreSamePublicECParameters(ECParameters aParameters, ECParameters bParameters)
```
This was pulled from `ECDsaCertificateExtensions` and changed to accept the `ECParameters`, and renamed to be clear this is only comparing the public properties of `ECParameters`.

vcsjones commented on src/libraries/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/X509Certificate2.cs:
```diff
@@ -837,7 +872,7 @@ public static X509Certificate2 CreateFromEncryptedPemFile(string certPemFilePath
         ///   The SubjectPublicKeyInfo from the certificate determines what PEM labels are accepted for the private key.
         ///   For RSA certificates, accepted private key PEM labels are "RSA PRIVATE KEY" and "PRIVATE KEY".
-        ///   For ECDSA certificates, accepted private key PEM labels are "EC PRIVATE KEY" and "PRIVATE KEY".
+        ///   For ECDSA and ECDH certificates, accepted private key PEM labels are "EC PRIVATE KEY" and "PRIVATE KEY".
```
Will XML doc edits eventually propogate back up to docs, or do I need to open a new PR after this one to modify existing docs?

bartonjs replied:
> Will XML doc edits eventually propogate back up to docs, or do I need to open a new PR after this one to modify existing docs?

Unfortunately, they don't propagate.  There's an import tool that imports the doc comments to the docs site for members which are entirely lacking documentation; but further changes happen in that repository and don't flow back to here... so updates here can't sensibly flow back to the docs site.

So, yeah, it'll need a manual PR to the docs site.  (I don't think it would need any sort of "starting in .NET 6.0" for this particular change)

vcsjones commented:
Hm. CNG is behaving differently between Win10 and Win8.1/7. I suppose this is related to CNG's strong opinion that EC keys are either ECDH or ECDSA.

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/CertificateExtensionsCommon.cs:
```diff
@@ -22,15 +22,30 @@ internal static class CertificateExtensionsCommon
             Oid algorithmOid = publicKey.Oid;
+
             if (oidValue != algorithmOid.Value)
                 return null;
```
Is this check still appropriate for ECDH?  Can't it use id-ecDH or id-ecPublicKey?

bartonjs replied:
(From a comment later, apparently not...)

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/X509Pal.cs:
```diff
@@ -83,6 +72,42 @@ private partial class AppleX509Pal : ManagedX509ExtensionProcessor, IX509Pal
+                        throw Interop.AppleCrypto.CreateExceptionForOSStatus(errSecUnsupportedKeySize);
+                    }
+
+                    return factory(key);
```
Since this is effectively a tail-call, would it make more sense to just have the key the return value, and avoid the callback and the generic method?

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509Encoder.cs:
```diff
@@ -14,13 +14,29 @@ namespace Internal.Cryptography.Pal
-        public AsymmetricAlgorithm DecodePublicKey(Oid oid, byte[] encodedKeyValue, byte[] encodedParameters, ICertificatePal? certificatePal)
+        public ECDsa DecodeECDsaPublicKey(ICertificatePal? certificatePal)
+        {
+            if (certificatePal is OpenSslX509CertificateReader reader)
```
Previously this was just a nullcheck then a hard cast.  Since the PAL type is internal it's... unlikely... that we'd have a second implementation in the same environment, but I wonder if the NotSupportedException if it happened would be harder to diagnose than the hard cast failing.

vcsjones replied:
Ah, the Apple one did it as `as Foo` and checked the result so I just had it stuck one way in my head. I think it makes sense to change both.

https://github.com/dotnet/runtime/blob/d0b699fd5d24f340b7a0ab4bf656f85ea9af822e/src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/X509Pal.cs#L31-L33

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/X509Pal.PublicKey.cs:
```diff
@@ -53,9 +74,13 @@ public AsymmetricAlgorithm DecodePublicKey(Oid oid, byte[] encodedKeyValue, byte
+                where TAlgorithm : AsymmetricAlgorithm, new()
         {
-            ECDsa ecdsa;
+            TAlgorithm ecdsa;
```
```suggestion
            TAlgorithm key;
```

(or something other than the name of one of the input algorithm types :smile:)

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/PublicKey.cs:
```diff
@@ -49,5 +71,69 @@ public AsymmetricAlgorithm Key
+                    throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding, e);
+                }
+
+                oid = new Oid(spki.Algorithm.Algorithm);
```
```suggestion
                oid = new Oid(spki.Algorithm.Algorithm, null);
```

To avoid the lookup to see if it was an OID value or a FriendlyName

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/X509Certificate2.cs:
```diff
@@ -926,12 +972,39 @@ public static X509Certificate2 CreateFromEncryptedPem(ReadOnlySpan<char> certPem
+                            password,
+                            ECDsa.Create,
+                            certificate.CopyWithPrivateKey),
+                    Oids.EcPublicKey when IsECDiffieHellman(certificate) =>
```
I think flipping these will fix the Windows problems. IIRC, ECDH keys can do ECDH or ECDSA, but ECDSA keys can't do ECDH.

bartonjs replied:
Is it even possible to get into this second EcPublicKey block on any OS?  For Linux and macOS it's just an ECC key.  For Windows I think the only thing that might stand a chance of failing this test is a Montgomery curve (Curve25519), since (IIRC) Montgomery curves are only allowed to do ECDH on Windows (I don't remember if that fails at import time, or signing time).

vcsjones replied:
The `GetECDsaPublicKey` and friends take in to account of the certificate's key usage. So if the cert has a keyUsage of only `keyAgreement`, `GetECDsaPublicKey` will return null and we'll end up in here.

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/tests/PublicKeyTests.cs:
```diff
@@ -558,6 +640,149 @@ public static void TestKey_BrainpoolP160r1()
+
+            PublicKey key = PublicKey.CreateFromSubjectPublicKeyInfo(spki, out int read);
+
+            Assert.True(key.Key is RSA, "is RSA");
```
Something like `Assert.IsAssignableTo<RSA>` would lead to a more easily diagnosed failure.

vcsjones committed:
Fix ECDH importing on downlevel Windows.

Windows 8.1 and below needs to be explicitly told that the public key it
is handling is for key agreement purposes.

vcsjones committed:
No FriendlyName lookup when constructing the OID.

vcsjones committed:
Use correct assertion for types.

vcsjones commented:
@bartonjs I think I fixed the Windows 7 / 8.1 key import problem. Does the change in 2d4fda5076435e90ccf8bfea22f481d3795ee4af make sense?

bartonjs commented:
>  Does the change in 2d4fda5 make sense?

Depends on your definition of "make sense" :smile:.  Apparently the private keys default to ECDH but the public keys were defaulting to ECDSA, and whatever lookup it did in the CryptFindOID universe returned multiple hits, ECDSA first... and that flag says "if something can't encrypt, I don't want it".. and while ECDH can't encrypt, it's purpose is eventually encryption, so it's marked as "encryption".  Does any of that make sense? Not especially :smile:.

Does the code read cleanly? Sure.  Does it make tests pass? Apparently :smile:.

vcsjones committed:
Simplify X509 PAL changes.

vcsjones committed:
XML docs for PublicKey.

vcsjones committed:
XML docs for X509Cert2 changes.

vcsjones committed:
Add Windows PFX loading tests

vcsjones committed:
Merge remote-tracking branch 'ms/master' into 413-impl

vcsjones committed:
Merge remote-tracking branch 'ms/master' into 413-impl

vcsjones committed:
Merge remote-tracking branch 'ms/master' into 413-impl

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509CertificateReader.cs:
```diff
@@ -632,6 +652,26 @@ public ICertificatePal CopyWithPrivateKey(ECDsa privateKey)
+            {
+                typedKey.ImportParameters(ecParameters);
+
+                return CopyWithPrivateKey((SafeEvpPKeyHandle)typedKey.DuplicateKeyHandle());
```
Why is this cast needed? That looks like the return type of DuplicateKeyHandle to me...

vcsjones replied:
It's not; this code was "inspired" from the ECDSA `CopyWithPrivateKey` which also has the redundant cast. I'll remove all of them.

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/X509Pal.PublicKey.cs:
```diff
@@ -53,10 +74,21 @@ public AsymmetricAlgorithm DecodePublicKey(Oid oid, byte[] encodedKeyValue, byte
+            TAlgorithm key;
+            CryptImportPublicKeyInfoFlags importFlags = CryptImportPublicKeyInfoFlags.NONE;
+
+            if (typeof(ECDiffieHellmanCng) == typeof(TAlgorithm))
```
We're doing strict equality on the type parameter, which means we have some expectations as to what's going on.

Is it worth adding a debugging aid, like

```C#
Debug.Assert(typeof(TAlgorithm).BaseType?.BaseType == typeof(AsymmetricAlgorithm), $"TAlgorithm ({typeof(TAlgorithm).FullName}) should be an implementation type, not an algorithm type.");
```

Sure, the assert would be wrong if we ever subclassed ECDiffieHellmanCng, but it turns out, this if would be wrong, too :smile:.

vcsjones replied:
I think I'd rather take a bool to this method and let the caller decide if this flag is applicable instead of trying to determine from the `TAlgorithm`. Something like `bool restrictPublicKeyToEncryption` on this method. The ECDSA paths will specify false and the ECDH path will say true. Would that be better?

bartonjs replied:
Either a bool, or (since it's already a Windows-PAL-only member) just the CryptImportPublicKeyInfoFlags themselves (perhaps defaulted to none) seems fine to me.

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/PublicKey.cs:
```diff
@@ -49,5 +87,103 @@ public AsymmetricAlgorithm Key
+
+        private AsnWriter EncodeSubjectPublicKeyInfo()
+        {
+            SubjectPublicKeyInfoAsn spki = new SubjectPublicKeyInfoAsn {
+                Algorithm = new AlgorithmIdentifierAsn {
```
```suggestion
            SubjectPublicKeyInfoAsn spki = new SubjectPublicKeyInfoAsn
            {
                Algorithm = new AlgorithmIdentifierAsn
                {
```

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/PublicKey.cs:
```diff
@@ -49,5 +87,103 @@ public AsymmetricAlgorithm Key
+            SubjectPublicKeyInfoAsn spki = new SubjectPublicKeyInfoAsn {
+                Algorithm = new AlgorithmIdentifierAsn {
+                    Algorithm = _oid.Value ?? string.Empty,
+                    Parameters = EncodedParameters.RawData
+                },
+                SubjectPublicKey = EncodedKeyValue.RawData
```
```suggestion
                    Parameters = EncodedParameters.RawData,
                },
                SubjectPublicKey = EncodedKeyValue.RawData,
```

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/X509Certificate2.cs:
```diff
@@ -686,6 +686,83 @@ public bool Verify()
+        /// A new ECDiffieHellman certificate with the <see cref="HasPrivateKey" /> property set to true.
+        /// The current certificate isn't modified.
+        /// </returns>
+        /// <exception cref="ArgumentNullException"><paramref name="privateKey" /> is null.</exception>
```
```suggestion
        /// <exception cref="ArgumentNullException"><paramref name="privateKey" /> is <see langword="null"/>.</exception>
```

vcsjones committed:
Remove redundant casts.

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/X509Certificate2.cs:
```diff
@@ -686,6 +686,83 @@ public bool Verify()
+        /// </summary>
+        /// <param name="privateKey">The private ECDiffieHellman key.</param>
+        /// <returns>
+        /// A new ECDiffieHellman certificate with the <see cref="HasPrivateKey" /> property set to true.
```
```suggestion
        /// A new ECDiffieHellman certificate with the <see cref="HasPrivateKey" /> property set to <see langword="true" />.
```

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/tests/CertificateCreation/CertificateRequestUsageTests.cs:
```diff
@@ -552,6 +552,42 @@ public static void InvalidSignatureAlgorithmEncoding(string caseName, string sig
+                ecdsa,
+                HashAlgorithmName.SHA256);
+
+                issuerRequest.CertificateExtensions.Add(
+                    new X509BasicConstraintsExtension(true, false, 0, true));
```
This statement seems to be indented wrong.

vcsjones committed:
Fix braces on new lines and add trailing commas.

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/tests/CertificateCreation/CertificateRequestUsageTests.cs:
```diff
@@ -552,6 +552,42 @@ public static void InvalidSignatureAlgorithmEncoding(string caseName, string sig
+            Assert.NotNull(publicCertKey);
+            Assert.Equal(ecdh.ExportSubjectPublicKeyInfo(), publicCertKey.ExportSubjectPublicKeyInfo());
+        }
+
```
Do you see value any of these extra tests:

* new CertificateRequest(ecdsa, ...) with KeyAgreement KU makes for a valid ECDH/invalid ECDSA cert
* new CertificateRequest(new PublicKey(ecdh), ...) without KeyAgreement makes the cert valid for ECDSA and ECDH
* Other sorts of "flexibility has increased" / "the output key can't be used to rerun this test".

The first bullet is actually interesting in that CreateSelfSigned will... probably... fail? Since CopyWithPrivateKey... should... fail?

vcsjones replied:
> CreateSelfSigned will... probably... fail? Since CopyWithPrivateKey... should... fail?

Yeah. It fails with

> The provided key does not match the public key algorithm for this certificate.

This is already existing behavior. I think the behavior is... reasonable, but maybe could be something we have additional validation on in `CreateSelfSigned` so the exception is clearer. If anything it seems not quite right that the result is an `ArgumentException`. I don't think folks should reasonably expect to create a cert with an ECDSA cert and get an ECDH back out.

vcsjones replied:
Actually we throw an `ArgumentException` in a lot of other cases for `Create`, so maybe it's closer to "okay" than I thought.

bartonjs replied:
> I don't think folks should reasonably expect to create a cert with an ECDSA cert and get an ECDH back out.

I agree, but if it works now (especially if it works in already shipped versions) we probably want to keep it working.

> but maybe could be something we have additional validation on in CreateSelfSigned so the exception is clearer.

I'm amenable to that (but it probably makes more sense as a followup change).

> Actually we throw an ArgumentException in a lot of other cases for Create, so maybe it's closer to "okay" than I thought.

We could (in a change other than this one) try to clean those up, too.

vcsjones replied:
>  if it works now (especially if it works in already shipped versions)

It does not nor did it. Opened #42751 to improve it later.

bartonjs replied:
Well, the thing I'm concerned works and could potentially not in the future is Create, not CreateSelfSigned.  (But it'd require us saving a lot more state than we currently do, so it's not super likely to break)

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/tests/PublicKeyTests.cs:
```diff
@@ -389,6 +450,27 @@ public static void TestECDsaPublicKey_ValidatesSignature()
+                otherParty.ImportFromPem(TestData.EcDhPkcs8Key);
+                byte[] key = otherParty.DeriveKeyFromHash(publicKey.PublicKey, HashAlgorithmName.SHA256);
+
+                byte[] expectedKey = new byte[] {
```
```suggestion
                byte[] expectedKey = new byte[]
                {
```

vcsjones committed:
Documentation feedback.

vcsjones committed:
Fix indentation

vcsjones committed:
Fix bracing on new line

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/tests/PublicKeyTests.cs:
```diff
@@ -558,6 +640,149 @@ public static void TestKey_BrainpoolP160r1()
+            Assert.Throws<CryptographicException>(() =>
+                PublicKey.CreateFromSubjectPublicKeyInfo(new byte[] { 0xFF }, out _));
+        }
+
```
How about: well formed, but not an algorithm we support, e.g. GOST R 34-10.2001.

```C#
// The SubjectPublicKeyInfo value from the test certificate in
// https://tools.ietf.org/html/rfc4491#section-4.2
const string KeyHex =
    "3063301C06062A8503020213301206072A85030202240006072A850302021E01" +
    "03430004408495687560021A407508CD138C31892CFDE505037A435CF46D2B0F" +
    "E74F327E578FEBCC16B9958803D09A7C85AE0FE48DEAA6BB7E56C7CBB0DF0F66" +
    "BCCAEA1A60";
```

Since it's just doing decode, it seems like it should crack it apart just fine, and put it back together just fine.


Additionally, it'll probably support cracking open and reassembling an invalid DSA key (malformed DSAParameters), but then the Key property will throw.  Probably another good test to have.

vcsjones committed:
Test PublicKey supports any SPKI.

PublicKey should at least be able to create a SPKI for any algorithm.

vcsjones committed:
Add more coverage for CertificateRequest.

Verify we can sign arbitrary SPKIs.

vcsjones committed:
Test invalid encoded key for known algorithm.

Add a test that we can still round-trip keys that contain invalid
parameters but are well structured.

vcsjones committed:
Pass in flags rather than determine from type parameter.

vcsjones committed:
Additional ECDSA test.

vcsjones commented:
@bartonjs thanks as always for the thoughtful feedback. I think I have addressed all of the points you have made so far.

vcsjones commented:
Guessing macOS is going to blow up with a bad DSA at usage time, not import time. No idea what what it is actually doing with an 8-bit DSA key.

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/X509Certificate2.cs:
```diff
@@ -724,14 +801,7 @@ public bool Verify()
-        /// </para>
-        /// <para>PEM-encoded items that have a different label are ignored.</para>
-        /// <para>
-        ///   Combined PEM-encoded certificates and keys do not require a specific order. For the certificate, the
+        ///   The Subjbool restrictPublicKeyToEncryption PEM-encoded certificates and keys do not require a specific order. For the certificate, the
         ///   the first certificate with a CERTIFICATE label is loaded. For the private key, the first private
```
This doesn't seem intentional...  Especially not "The subjbool restrictPublicKeyToEncryption"

vcsjones replied:
... 
![Doh](https://user-images.githubusercontent.com/361677/94321725-713c3000-ff5e-11ea-9db9-d34f9e482112.png)

bartonjs commented:
> Guessing macOS is going to blow up with a bad DSA at usage time, not import time. No idea what what it is actually doing with an 8-bit DSA key.

Well, p is 0, so that's a 0-bit key, no? :smile:.

I'd probably delete one of the integers out of the DSA-Params struct so it doesn't decode correctly, making the whole "bad DSA" SPKI

```
30 1B 30 13 06 07 2A 86 48 CE 38 04 01 30 08 02 
01 00 02 03 00 FF FF 03 04 00 02 01 03 
```

vcsjones committed:
Fix inadvertent changes

vcsjones committed:
Use a more broken DSA key.

bartonjs reviewed (approved):
Once again: Thanks, @vcsjones!
